crates/apollo-parser/src/ast/node_ext.rs

```diff
@@ -193,12 +196,164 @@ fn unescape_string(input: &str) -> String {
+struct GraphQLLines<'a> {
+    input: &'a str,
+    finished: bool,
+}
+
+impl<'a> GraphQLLines<'a> {
+    fn new(input: &'a str) -> Self {
+        Self {
+            input,
+            finished: false,
+        }
+    }
+}
+
+impl<'a> Iterator for GraphQLLines<'a> {
+    type Item = &'a str;
+    fn next(&mut self) -> Option<Self::Item> {
+        // Can't just check for the input string being empty, as an empty string should still
+        // produce one line.
+        if self.finished {
+            return None;
+        }
+
+        let Some(index) = memchr::memchr2(b'\r', b'\n', self.input.as_bytes()) else {
+            self.finished = true;
+            return Some(self.input);
+        };
+        let line = &self.input[..index];
+        let rest = match self.input.get(index..=index + 1) {
+            Some("\r\n") => &self.input[index + 2..],
+            _ => &self.input[index + 1..],
+        };
+        self.input = rest;
+        Some(line)
+    }
+}
+
+/// Split lines on \n, \r\n, and just \r
+fn split_lines(input: &str) -> impl Iterator<Item = &str> {
+    GraphQLLines::new(input)
+}
+
+/// Replace a literal pattern in a string but push the output to an existing string.
+///
+/// Like `str::replace`, but doesn't allocate if there's enough space in the provided output.
+fn replace_into(input: &str, pattern: &str, replace: &str, output: &mut String) {
+    let mut last_index = 0;
+    for index in memchr::memmem::find_iter(input.as_bytes(), pattern.as_bytes()) {
+        output.push_str(&input[last_index..index]);
+        output.push_str(replace);
+        last_index = index + pattern.len();
+    }
+    if last_index < input.len() {
+        output.push_str(&input[last_index..]);
+    }
+}
+
+/// Implementation of the spec function `BlockStringValue(rawValue)`. In addition to handling
+/// indents and newline normalization, this also handles escape sequences (strictly not part of
+/// BlockStringValue in the spec, but more efficient to do it at the same time).
+///
+/// Spec: https://spec.graphql.org/October2021/#BlockStringValue()
+fn unescape_block_string(raw_value: &str) -> String {
+    /// WhiteSpace :: Horizontal Tab (U+0009) Space (U+0020)
+    fn is_whitespace(c: char) -> bool {
+        matches!(c, ' ' | '\t')
+    }
+    /// Check if a string is all WhiteSpace. This expects a single line of input.
+    fn is_whitespace_line(line: &str) -> bool {
+        line.chars().all(is_whitespace)
+    }
+    /// Count the indentation of a single line (how many WhiteSpace characters are at the start).
+    fn count_indent(line: &str) -> usize {
+        line.chars().take_while(|&c| is_whitespace(c)).count()
+    }
+
+    // 3. For each line in lines:
+    let common_indent = split_lines(raw_value)
+        // 3.a. If line is the first item in lines, continue to the next line.
+        .skip(1)
+        .filter_map(|line| {
+            // 3.b. Let length be the number of characters in line.
+            let length = line.len();
```

Comment:
Please add a comment to explain that while `str.len()` counts UTF-8 bytes and `count_indent` counts `char`s, comparing them is still ok because the WhiteSpace characters we’re looking for are in the ASCII range and so occupy one UTF-8 byte each.